Support dual source blending

freqmod opened the pull request:
Support writing shaders with alt blender mode.
Tested on linux with OpenGL, and Vulkan, and on Mac using metal. I have made some changes for Direct X but since I don't have a windows build environment I haven't tested that.

See https://github.com/freqmod/helicoid/tree/master/helicoid-wgpu for usage example.
See https://github.com/gfx-rs/wgpu/issues/1804#issuecomment-1669452915 for related issue

freqmod commented:
I had a look into the clippy errors, and it seems to be in some kind of external crate that I have not changed.

ErichDonGubler commented on src/back/msl/mod.rs:
```diff
@@ -82,7 +82,7 @@ pub type EntryPointResourceMap = std::collections::BTreeMap<String, EntryPointRe
     BuiltIn(crate::BuiltIn),
     Attribute(u32),
-    Color(u32),
+    Color(u32, Option<u32>),
```
nitpick: I don't think it's obvious what these fields mean any more; named fields are probably a better fit here.

ErichDonGubler reviewed (commented):
I can't leave a full review, since I'm not familiar with the domain and don't trust myself to validate shader output properly (CC @teoxoy), but most of the changes in code don't look controversial to me, assuming the shader output is correct. In addition to spanned commentary, here are some reviewer notes I made for myself:

* Original motivating issue in `wgpu`: https://github.com/gfx-rs/wgpu/issues/1804
* The [`Dual Source Blending` section of the `Blending` article from Khronos' wiki](https://www.khronos.org/opengl/wiki/Blending#Dual_Source_Blending) was particularly useful for helping me understand the domain here. This source also corroborates the GLSL logic added here, so yay!
* This adds support with these ends:
	* Front end: only WGSL
	* Back ends: all supported backends.
		* Q: Do we have sources for these syntaxes?
	    * **@freqmod mentions adding support for DirectX, which I presume to mean HLSL, but I don't see the HLSL represented here.** I found https://github.com/KhronosGroup/SPIRV-Cross/pull/815, which looks like a potentially helpful model for what to do here (viz., use `SV_Target${index}`).

ErichDonGubler commented on src/back/glsl/mod.rs:
```diff
@@ -333,6 +333,18 @@ struct VaryingName<'a> {
+                    // fragment to pipeline
+                    (ShaderStage::Fragment, true) => "fs2p",
+                };
+                write!(f, "_{prefix}_location{index}",)
```
nitpick: Fun fact, `rustfmt` doesn't remove trailing commas from macros, even when not needed for macros like `write!(…)`, because nobody's done the work of special-casing for "known" macros where it's okay yet: https://github.com/rust-lang/rustfmt/issues/3065

freqmod commented:
> @freqmod mentions adding support for DirectX, which I presume to mean HLSL, but I don't see the HLSL represented here. I found https://github.com/KhronosGroup/SPIRV-Cross/pull/815, which looks like a potentially helpful model for what to do here (viz., use SV_Target${index}).

Regarding HLSL it may be that i only did some work on that in the WGPU patch (and not the naga patch) and mixed up.

Regarding example code: there is a shader in https://github.com/freqmod/helicoid/blob/master/helicoid-gpurender/shaders/text.fs.wgsl but it may be useful to get one in this (or the wgpu) repo for proper regression testing.

ErichDonGubler commented:
We should definitely have shader snapshot tests for this here, independent of `wgpu`. That would make the intent of the output of these changes clearer.

freqmod commented:
- I did an attempt at some HLSL code based on [the](https://github.com/KhronosGroup/SPIRV-Cross/pull/815) as suggested  (still I haven't set up a windows dev environment).
- I have added a test for dual source blending. I got some problems with the validation of bindings that I have hacked around, although that should probably be done more properly with some input from the maintainers of that code.
- I also fixed the small code comments I got.

freqmod commented on src/front/wgsl/parse/mod.rs:
```diff
@@ -142,6 +143,11 @@ impl BindingParser {
                 self.location = Some(Parser::non_negative_i32_literal(lexer)?);
                 lexer.expect(Token::Paren(')'))?;
             }
+            "index" => {
+                lexer.expect(Token::Paren('('))?;
+                self.index = Some(Parser::non_negative_i32_literal(lexer)?);
+                lexer.expect(Token::Paren(')'))?;
+            }
```
- would you just have an attribute called bind_src_1? like this?
```
struct FragmentOutput{
    @location(0) color: vec4<f32>,
    @location(0) @bind_src_1 mask: vec4<f32>,
};
```
Ideally it would be nice to keep to whatever the wgpu spec https://www.w3.org/TR/WGSL/ has standardized but i don't know how the procedure to get some changes there is. It is probably too cumbersome, so it is better to just stay proprietary for now.

teoxoy replied:
> * would you just have an attribute called bind_src_1? like this?
> 
> ```
> struct FragmentOutput{
>     @location(0) color: vec4<f32>,
>     @location(0) @bind_src_1 mask: vec4<f32>,
> };
> ```

Yes, but `blend_src_1` not bind.

> Ideally it would be nice to keep to whatever the wgpu spec https://www.w3.org/TR/WGSL/ has standardized but i don't know how the procedure to get some changes there is. It is probably too cumbersome, so it is better to just stay proprietary for now.

I will advocate for `blend_src_1` but if it gets standardized as something else we can change our implementation.

> Would you like to translate @bind_src_1 to index 1 internally, or would you remove the index parameter on location in the generic data structure and rather have a bind_src_1 bool?

Yeah, I think that makes the most sense.

freqmod replied:
I used alternate_blend_source for the internal variable, as i think that reflects better what it is, although if you insist i can do a rename of the variables. I used blend_src_1 for the wgsl tough as it reflects the SRC1 when setting up the blend function.

freqmod commented on src/back/spv/writer.rs:
```diff
@@ -1473,6 +1474,13 @@ impl Writer {
                     }
                 }
+                if let Some(index) = index {
+                    debug_assert!(
+                        class == spirv::StorageClass::Output
+                            && stage == crate::ShaderStage::Fragment,
+                    );
```
ok, i added it as it was an if test in the branch below, but i guess it may not be the right place to do it as a validation.

teoxoy commented on src/back/glsl/features.rs:
```diff
@@ -233,6 +236,10 @@ impl FeaturesManager {
             // https://www.khronos.org/registry/OpenGL/extensions/ARB/ARB_texture_query_levels.txt
             writeln!(out, "#extension GL_ARB_texture_query_levels : require")?;
         }
+        if self.0.contains(Features::DUAL_SOURCE_BLENDING) {
```
```suggestion
        if self.0.contains(Features::DUAL_SOURCE_BLENDING) && version.is_es() {
```

teoxoy commented on tests/snapshots.rs:
```diff
@@ -619,11 +629,33 @@ fn convert_wgsl() {
+                Err(e) => panic!("{}", e.emit_to_string(&file)),
+            }
+        }
+    }
```
Let's specify `god_mode` in the `.param.ron` file of this snapshot to keep things consistent.

I opened https://github.com/gfx-rs/naga/issues/2442 to track passing finer grained `Capabilities`.

freqmod commented:
I accepted the proposed changes that i think are relevant. I will do another upload with manual changes to address the other comments later.

freqmod commented:
I would still prefer to have something like secondary_blend_source etc, but since you are more likely to maintain this in the future i think it is better to just go with your name and get this in without further bikeshedding.

freqmod commented:
Pushed rebase. It should not be any changes (except adaptions to upstream) since last push.

teoxoy commented:
> I would still prefer to have something like `secondary_blend_source`

How about `second_blend_source`?

I can make the change.

Otherwise, PR looks great! Thanks for all the changes :)

freqmod commented:
second_blend_source is good. I can do an upload. I'll also squash everything as the commits now are quite messy..

teoxoy commented:
If you are at it, can you update the WGSL attribute and also I saw a few remaining places (esp. errors) where we still use "alternate". Thanks!

freqmod commented:
I will review the squashed version before pushing to fix any small things left.

freqmod commented:
Now I have addressed everything I can see. Feel free to do any changes on top of the last upload you need and merge when you are happy. Thanks for working on the review.

teoxoy commented:
Thanks! I pushed a few more commits, let me know if they look good and we can land it!

freqmod commented:
Thanks. They look good to me, please land it.
